### PR TITLE
storage: Add a "node is alive" metric

### LINKED ISSUE #12222: storage: add "live nodes" metric
The node liveness table gives us a good signal within a process about the live nodes in the cluster. We should have a single node within the cluster export either a count of the live nodes, or a per-node "live" metric. Possible choices for this node are the leaseholder of range 1 (i.e. the first range) or the smallest "live" node. The "live nodes" metric would be used to power a "nodes" graph similar to what we display on our internal grafana graphs using the prometheus "up" metric.

Cc @mrtracy

### PR DESCRIPTION
**Pull request by a-robinson**
The idea being that you would aggregate this metric across your nodes to
determine how many healthy nodes you have at any given time, while still
maintaining the ability to check each individual node's health.

For this to graph well in the admin UI, we'll need to have a pretty low leniency for missing data points (given that a non-live node is unlikely to be able to persist its timeseries data).

Fixes #12222

@mrtracy

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/12296)

### CODE REVIEW

**Comment by petermattis** [comment]
A down or unhealthy node won't be able to record this metric during the normal periodic metric recording. See `ts.DB.PollSource` and `server/status.MetricsRecorder`. I think that will make this metric somewhat less useful for our admin UI. Being able to record the metric was the motivation behind using a well known live node in the cluster as described in #12222. But perhaps I'm over-thinking this and the current solution is ok if we adjust querying for this metric so that non-existent data indicates the metric is 0. 

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/storage/node_liveness.go, line 116 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12296#-KYpiaf6mCWV9pqj5Kvu:-KYpiaf6mCWV9pqj5Kvv:b-b1wh4o) ([raw file](https://github.com/cockroachdb/cockroach/blob/8536c1e13885d722a8e3f246a884e59c875db5c4/pkg/storage/node_liveness.go#L116)):*
> ```Go
> 		heartbeatInterval: heartbeatInterval,
> 		metrics: LivenessMetrics{
> 			NodeIsAlive:        metric.NewGauge(metaNodeIsAlive),
> ```

I saw that our underlying metrics library has a `FunctionalGauge` type that used a callback to get the value. This would be a good use case.

---

*[pkg/storage/node_liveness_test.go, line 118 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12296#-KYpgn36LI2byWXBan4R:-KYpgn36LI2byWXBan4S:b-uetdu2) ([raw file](https://github.com/cockroachdb/cockroach/blob/8536c1e13885d722a8e3f246a884e59c875db5c4/pkg/storage/node_liveness_test.go#L118)):*
> ```Go
> func TestNodeLivenessEpochIncrement(t *testing.T) {
> 	defer leaktest.AfterTest(t)()
> 	storage.SelfLivenessCheckInterval = time.Millisecond
> ```

You're not resetting this global at the end of the test which means the setting will leak over to other tests. You can either make this some sort of configuration value on `multiTestContext` or do something like:

```
defer func(d time.Duration) {
  storage.SelfLivenessCheckInterval = d
}(storage.SelfLivenessCheckInterval)
SelfLivenessCheckInterval = time.Millisecond
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12296#-:-KYphC3Rvr-pP-Aon2Pn:b-y5c5cg)*

**Comment by a-robinson** [comment]
@mrtracy, up for talking today about @petermattis's comment above?

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/storage/node_liveness.go, line 116 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12296#-KYpiaf6mCWV9pqj5Kvu:-KYsrHpbarRWAbFCID10:b-w7uv8o) ([raw file](https://github.com/cockroachdb/cockroach/blob/8536c1e13885d722a8e3f246a884e59c875db5c4/pkg/storage/node_liveness.go#L116)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I saw that our underlying metrics library has a `FunctionalGauge` type that used a callback to get the value. This would be a good use case.
</blockquote></details>

Aha, thanks for finding that. I looked for something along those lines but wasn't seeing it. Although it reopens the mess around our Gauge wrapper that came up in #12106. I'm hoping they'll let me add Inc and Dec methods to the Gauge interface upstream, although we'll see.

I'll fix this after chatting with @mrtracy.

---

*[pkg/storage/node_liveness_test.go, line 118 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12296#-KYpgn36LI2byWXBan4R:-KYsrN64Zklhl28AyAZt:b-lhnq79) ([raw file](https://github.com/cockroachdb/cockroach/blob/8536c1e13885d722a8e3f246a884e59c875db5c4/pkg/storage/node_liveness_test.go#L118)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

You're not resetting this global at the end of the test which means the setting will leak over to other tests. You can either make this some sort of configuration value on `multiTestContext` or do something like:

```
defer func(d time.Duration) {
  storage.SelfLivenessCheckInterval = d
}(storage.SelfLivenessCheckInterval)
SelfLivenessCheckInterval = time.Millisecond
```
</blockquote></details>

Good catch, but no longer relevant after switching to the FunctionalGauge.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12296#-:-KYsu_G52T_G0yxXStar:bjewfj6)*

**Comment by a-robinson** [comment]
Changed to only export the metric from the lowest-ID node that's alive. PTAL

---

Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/storage/node_liveness.go, line 116 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12296#-KYpiaf6mCWV9pqj5Kvu:-KYyk7rfG3uxcE0uT0R_:b-p330ei) ([raw file](https://github.com/cockroachdb/cockroach/blob/8536c1e13885d722a8e3f246a884e59c875db5c4/pkg/storage/node_liveness.go#L116)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Aha, thanks for finding that. I looked for something along those lines but wasn't seeing it. Although it reopens the mess around our Gauge wrapper that came up in #12106. I'm hoping they'll let me add Inc and Dec methods to the Gauge interface upstream, although we'll see.

I'll fix this after chatting with @mrtracy.
</blockquote></details>

Switched.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12296#-:-KYykA3Pj4JZohrg5vBz:buo2vet)*

**Comment by petermattis** [comment]
:lgtm:

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12296#-:-KYytWCu8yE8x7jTSGB-:bnfp4nl)*

**Comment by tamird** [comment]
Reviewed 4 of 4 files at r2, 1 of 2 files at r3.
Review status: 3 of 4 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/storage/node_liveness.go, line 113 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12296#-KYyueCGWyn72x7jXu_p:-KYyueCGWyn72x7jXu_q:b-9k4r3g) ([raw file](https://github.com/cockroachdb/cockroach/blob/a16027ade13bad3430a5486ac89362554233f4e0/pkg/storage/node_liveness.go#L113)):*
> ```Go
> 	}
> 	nl.metrics = LivenessMetrics{
> 		LiveNodes: metric.NewFunctionalGauge(metaLiveNodes, func() int64 {
> ```

metric.NewFunctionalGauge(metaLiveNodes, nl.numLiveNodes)`?

---

*[pkg/storage/node_liveness.go, line 403 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12296#-KYyv5uFj8FJe88OrVys:-KYyv5uFj8FJe88OrVyt:b-snmmxf) ([raw file](https://github.com/cockroachdb/cockroach/blob/a16027ade13bad3430a5486ac89362554233f4e0/pkg/storage/node_liveness.go#L403)):*
> ```Go
> func (nl *NodeLiveness) numLiveNodes() int64 {
> 	selfID := nl.gossip.NodeID.Get()
> 	livenessMap := nl.GetLivenessMap()
> ```

this does a full iteration over the internal map (under lock), and then we iterate over this map again below. Probably better to not go through this getter in this case and just do this whole method under lock.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12296)*

**Comment by a-robinson** [comment]
Review status: 3 of 4 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/storage/node_liveness.go, line 113 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12296#-KYyueCGWyn72x7jXu_p:-KYzE86A20_X8Sq3IrSV:b-5p5vq5) ([raw file](https://github.com/cockroachdb/cockroach/blob/a16027ade13bad3430a5486ac89362554233f4e0/pkg/storage/node_liveness.go#L113)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

metric.NewFunctionalGauge(metaLiveNodes, nl.numLiveNodes)`?
</blockquote></details>

Done. Silly C++ brain

---

*[pkg/storage/node_liveness.go, line 403 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12296#-KYyv5uFj8FJe88OrVys:-KYzJI-SL4CfQt8pR1cH:b-acl2fo) ([raw file](https://github.com/cockroachdb/cockroach/blob/a16027ade13bad3430a5486ac89362554233f4e0/pkg/storage/node_liveness.go#L403)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this does a full iteration over the internal map (under lock), and then we iterate over this map again below. Probably better to not go through this getter in this case and just do this whole method under lock.
</blockquote></details>

I wrote it this way assuming the overhead would be negligible and keeping the code simple is better. If you're interested in a more optimized version, how about this?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12296)*

File: pkg/storage/node_liveness.go [hunk 1]
```diff
@@ -48,6 +48,7 @@ var (
 
 // Node liveness metrics counter names.
 var (
+	metaLiveNodes          = metric.Metadata{Name: "liveness.livenodes"}
```
Comment (mberhault) [inline pkg/storage/node_liveness.go]: can you add `Help` fields to each of those metrics describing what they are. See `/_status/vars` for various examples.

**Review by mberhault — APPROVED** [review]

**Comment by tamird** [comment]
discussed in person - since node liveness is maintained via gossip and may be stale (leading to under or over counting), we should have every node (which believes itself to be live) count all live nodes, and then take the maximum value for display in the admin UI and grafana.

---

Review status: 2 of 4 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12296#-:-KYzNQWWCqUs2SXcEBTu:bujygvc)*

**Comment by a-robinson** [comment]
Updated as mentioned by @tamird

---

Review status: 2 of 4 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/storage/node_liveness.go, line 51 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12296#-KYp6iwC2ZO82tkM-oOa-r4-51:-KYzPwgkbaYheG8TMdc_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b884ee904457b061402d32a304d8d4d8af015aed/pkg/storage/node_liveness.go#L51)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

can you add `Help` fields to each of those metrics describing what they are. See `/_status/vars` for various examples.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12296#-:-KYzPs2v3ngPyNAelgEV:b-cea4pf)*

**Comment by tamird** [comment]
Reviewed 1 of 2 files at r5.
Review status: 3 of 4 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12296)*

**Comment by tamird** [comment]
Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12296)*

**Comment by a-robinson** [comment]
@mrtracy - where do we want to add the graph to the admin UI? I can add it in, but it doesn't have an obvious home.

**Comment by petermattis** [comment]
Review status: 2 of 4 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/storage/node_liveness.go, line 417 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/12296#-KZ-C2lwrCEyzU_Thcc3:-KZ-C2lxEq6GHWyrO3PT:bj2vhwu) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8f00a61e25f7543aab35c1b4970914b35d5ba76/pkg/storage/node_liveness.go#L417)):*
> ```Go
> // numLiveNodes is used to populate a metric that tracks the number of live
> // nodes in the cluster. Returns 0 if this node is not itself live, to avoid
> // reporting potentially inaccurate data.
> ```

Can you expand this comment to include the discussion of why every live node reports this metric vs only a particular node? It sounds like you and @tamird had some rationale which isn't captured here and probably won't be obvious to our future selves.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12296)*

**Comment by a-robinson** [comment]
Review status: 2 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/storage/node_liveness.go, line 417 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/12296#-KZ-C2lwrCEyzU_Thcc3:-KZ24eLv1Eh9tgwdmA6U:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8f00a61e25f7543aab35c1b4970914b35d5ba76/pkg/storage/node_liveness.go#L417)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Can you expand this comment to include the discussion of why every live node reports this metric vs only a particular node? It sounds like you and @tamird had some rationale which isn't captured here and probably won't be obvious to our future selves.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12296)*

**Comment by tamird** [comment]
Reviewed 2 of 2 files at r6, 1 of 2 files at r7, 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12296)*

**Comment by mrtracy** [comment]
@a-robinson don't bother adding the graph right now.  Normally, you would add it to an appropriate section of "nodeGraphs.tsx" in our ui code, but I am currently in the process of leveling that code - basically changing all of the graphs being displayed to better align with our needs.  I will take care of adding this node liveness graph at that time, it's a big ask and you've done the tough work already.

**Comment by a-robinson** [comment]
@tamird, any final comments before merge?

**Comment by tamird** [comment]
nope, go for it

**Pull request merged**